### PR TITLE
Change exception type for incorrect SVG date metadata

### PR DESCRIPTION
**Pull request by oscargus**
## PR Summary

Better checking of metadata with better error messages. Tests for all metadata errors (changed ValueError to TypeError as should be more appropriate, release note required?). 

Test for `_escape_attrib`.

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/tests/test_backend_svg.py [hunk 1]
```diff
@@ -469,3 +469,18 @@ def test_svg_metadata():
     values = [node.text for node in
               rdf.findall(f'./{CCNS}Work/{DCNS}subject/{RDFNS}Bag/{RDFNS}li')]
     assert values == metadata['Keywords']
+
+
+@pytest.mark.parametrize('metadata,error,message', [
+    ({'Date': 1}, TypeError, "Invalid type for Date metadata. Expected str"),
+    ({'Date': [1]}, TypeError, "Invalid type for Date metadata. Expected it"),
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_backend_svg.py:476]:
```suggestion
    ({'Date': [1]}, TypeError, "Invalid type for Date metadata. Expected iterable"),
```
to make it more readable. Break the line if necessary.

File: lib/matplotlib/tests/test_backend_svg.py [hunk 2]
```diff
@@ -469,3 +469,18 @@ def test_svg_metadata():
...
+    ({'Keywords': 1}, TypeError,
+     "Invalid type for Keywords metadata. Expected str"),
+    ({'Keywords': [1]}, TypeError,
+     "Invalid type for Keywords metadata. Expected it"),
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_backend_svg.py:480]:
```suggestion
     "Invalid type for Keywords metadata. Expected iterable"),
```

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/backends/backend_svg.py [hunk 3]
```diff
@@ -481,14 +490,6 @@ def _make_id(self, type, content):
     def _make_flip_transform(self, transform):
         return transform + Affine2D().scale(1, -1).translate(0, self.height)
 
-    def _get_font(self, prop):
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_svg.py:484]: This is pretty unrelated; it should be a separate commit at least.

Comment (oscargus) [reply lib/matplotlib/backends/backend_svg.py]: I can see why it looks like that. ;-) (Originally, the purpose of the PR was to increase test coverage.)

I've moved it to a separate PR.

**Comment by oscargus** [comment]
The linting failures are not correct as the code runs...

File: lib/matplotlib/backends/backend_svg.py [hunk 4]
```diff
@@ -420,8 +425,12 @@ def ensure_metadata(mid):
...
             if info is not None:
                 mid = ensure_metadata(mid)
-                writer.element(f'dc:{key}', text=info)
-
+                if isinstance(info, str):
+                    writer.element(f'dc:{key}', text=info)
+                else:
+                    raise TypeError(f'Invalid type for {key.title()} '
+                                    f'metadata. Expected str'
+                                    f', not {type(info)}.')
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_svg.py:433]: And break after the comma, not before.

**Review by QuLogic — COMMENTED** [review]
There seems to be a mix of `.format` and f-strings, and I don't see a reason why they should differ.

**Comment by oscargus** [comment]
I guess that was a copy-and-paste thing, where I changed at some stage. Now it is refactored and uses some common check-and-error functions to avoid code duplication. (Didn't find it worthwhile to make it flexible enough to also deal with the different date formats.)

File: lib/matplotlib/backends/backend_svg.py [hunk 5]
```diff
@@ -298,6 +298,24 @@ def generate_css(attrib=None):
...
+def _check_is_iterable_of_str(infos, key):
+    for info in infos:
+        if not isinstance(info, str):
+            raise TypeError(f'Invalid type for {key} metadata. Expected '
+                            f'iterable of str, not {type(info)}.')
+
+
+def _raise_not_str_or_iterable_of_str(infos, key):
+    raise TypeError(f'Invalid type for {key} metadata. Expected str or '
+                    f'iterable of str, not {type(infos)}.')
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_svg.py:316]: These two could be joined, I think? One or the other is called based on `np..iterable`, which could be in a single function here.

Comment (oscargus) [reply lib/matplotlib/backends/backend_svg.py:316]: You are correct. Well spotted!

**Review by QuLogic — APPROVED** [review]
Can you update the commit message to match the PR title?

**Comment by oscargus** [comment]
> Can you update the commit message to match the PR title?

Sure. Is the general rule that one should include that much info? I did it on purpose as saying SVG in a commit affecting backend_svg seems a bit bloated...

**Comment by QuLogic** [comment]
I dunno, "Better error messages" sounds too vague to me, while "Fix some typos" is vague enough to not need specifics...

File: lib/matplotlib/backends/backend_svg.py [hunk 6]
```diff
@@ -376,14 +395,14 @@ def _write_metadata(self, metadata):
                     elif isinstance(d, (datetime.datetime, datetime.date)):
                         dates.append(d.isoformat())
                     else:
-                        raise ValueError(
-                            'Invalid type for Date metadata. '
-                            'Expected iterable of str, date, or datetime, '
-                            'not {!r}.'.format(type(d)))
+                        raise TypeError(
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_svg.py:398]: This needs an API change note and it is changing the exception type on public API.

File: lib/matplotlib/tests/test_backend_svg.py [hunk 7]
```diff
@@ -464,3 +465,37 @@ def test_svg_metadata():
...
+def test_svg_incorrect_metadata(metadata, error, message):
+    with pytest.raises(error, match=message), BytesIO() as fd:
+        fig, ax = plt.subplots()
+        fig.savefig(fd, format='svg', metadata=metadata)
+
+
+def test_svg_escape():
+    fig, ax = plt.subplots()
+    fig.text(0.5, 0.5, "<\'\"&>", gid="<\'\"&>")
+    ax.set_axis_off()
```
Comment (anntzer) [inline lib/matplotlib/tests/test_backend_svg.py:497]: Not that it matters much, but you can do `fig = plt.figure(); fig.text(...)` (not having an axes at all) instead of hiding the axes.

**Review by timhoffm — APPROVED** [review]
